web-common/src/features/dashboards/stores/AdvancedMeasureCorrector.ts

```diff
@@ -0,0 +1,153 @@
+  private measuresGrains: Map<string, V1TimeGrain>;
+
+  private constructor(
+    private readonly dashboard: MetricsExplorerEntity,
+    private readonly metricsViewSpec: V1MetricsViewSpec,
+  ) {
+    this.measuresMap = getMapFromArray(
+      metricsViewSpec.measures ?? [],
+      (m) => m.name ?? "",
+    );
+    this.measuresGrains = getMapFromArray(
+      metricsViewSpec.measures ?? [],
+      (m) => m.name ?? "",
+      (m) => {
+        const d = m.requiredDimensions?.find(
+          (d) =>
+            d.timeGrain && d.timeGrain !== V1TimeGrain.TIME_GRAIN_UNSPECIFIED,
+        );
+        return d?.timeGrain ?? V1TimeGrain.TIME_GRAIN_UNSPECIFIED;
+      },
+    );
+  }
+
+  public static correct(
+    dashboard: MetricsExplorerEntity,
+    metricsViewSpec: V1MetricsViewSpec,
+  ) {
+    new AdvancedMeasureCorrector(dashboard, metricsViewSpec).correct();
+  }
+
+  private correct() {
+    this.correctFilters();
+    this.correctLeaderboards();
+    this.correctTimeDimensionDetails();
+    this.correctPivot();
+  }
+
+  private correctFilters() {
+    this.dashboard.dimensionThresholdFilters.forEach((dimensionThreshold) => {
+      dimensionThreshold.filter = filterIdentifiers(
+        dimensionThreshold.filter,
+        (_, ident) => !this.measureMismatch(ident, false, false),
+      ) as V1Expression;
+    });
+    this.dashboard.dimensionThresholdFilters =
+      this.dashboard.dimensionThresholdFilters.filter(
+        (dt) => dt.filter !== undefined,
+      );
+  }
+
+  private correctLeaderboards() {
+    if (
+      this.dashboard.leaderboardMeasureName &&
+      !this.measureMismatch(this.dashboard.leaderboardMeasureName, true, false)
+    ) {
+      return;
+    }
+
+    this.dashboard.leaderboardMeasureName = "";
+    for (const measure of this.metricsViewSpec.measures ?? []) {
+      if (!this.measureMismatch(measure.name ?? "", true, false)) {
+        this.dashboard.leaderboardMeasureName = measure.name ?? "";
+        break;
+      }
+    }
+  }
+
+  private correctTimeDimensionDetails() {
+    if (
+      !this.measureMismatch(
+        this.dashboard.tdd.expandedMeasureName ?? "",
+        true,
+        false,
+      )
+    ) {
+      return;
+    }
+
+    this.dashboard.tdd.expandedMeasureName = "";
+    if (
+      this.dashboard.activePage ===
+      DashboardState_ActivePage.TIME_DIMENSIONAL_DETAIL
+    ) {
+      this.dashboard.activePage = DashboardState_ActivePage.DEFAULT;
+    }
+  }
+
+  private correctPivot() {
+    this.dashboard.pivot.columns.measure =
+      this.dashboard.pivot.columns.measure.filter(
+        (m) => !this.measureMismatch(m.id, true, false),
+      );
+    this.dashboard.pivot.sorting = this.dashboard.pivot.sorting.filter(
+      (s) =>
+        !this.measuresMap.has(s.id) || !this.measureMismatch(s.id, true, false),
+    );
+  }
+
+  private measureMismatch(
+    measureName: string,
+    supportsComparisonMeasure: boolean,
+    supportsWindowedMeasure: boolean,
+  ) {
```

Comment:
This function signature is hard to understand. Mind making the function name more descriptive? And maybe add a function-level comment?